slatedb/src/cached_object_store/storage_fs.rs

```diff
@@ -580,30 +578,34 @@ impl FsCacheEvictorInner {
             .object_store_cache_bytes
             .set(self.cache_size_bytes.load(Ordering::Relaxed));
 
-        if !evict {
-            return 0;
-        }
-
         // if the cache size is still below the limit, do nothing
         if self.cache_size_bytes.load(Ordering::Relaxed) <= self.max_cache_size_bytes as u64 {
             return 0;
         }
         // TODO: check the disk space ratio here, if the disk space is low, also triggers evict.
 
-        // if the cache size exceeds the limit, evict the cache files in batch with the batch_factor,
-        // this may help to avoid the cases like triggering the evictor too frequently when the cache
-        // size is just slightly above the limit.
-        let mut total_bytes: usize = 0;
-        for _ in 0..self.batch_factor {
-            let evicted_bytes = self.maybe_evict_once().await;
-            if evicted_bytes == 0 {
-                return total_bytes;
+        // The maximum byte size the cache will take up on disk. If a write would cause the
+        // cache to exceed this threshold, entries are evicted using an 2-random strategy until
+        // the cache reaches 90% of `max_cache_size_bytes`.
+        //
+        // It's ok to call evict after inserting the new entry, because we will evict entries with eailer `accessed_time`.
+        // This ensures that the newly added entry will not be evicted immediately.
+        let mut evicted_bytes: usize = 0;
+        if evict && self.cache_size_bytes.load(Ordering::Relaxed) > self.max_cache_size_bytes as u64
+        {
+            // We sacrifice floating-point precision error to prevent possible overflow(i.e. self.max_cache_size_bytes * 9 / 10).
+            let target_size = ((self.max_cache_size_bytes as f64) * 0.9) as u64;
+            while self.cache_size_bytes.load(Ordering::Relaxed) > target_size {
+                // TODO(asukamilet): reduce the number of lock acquisitions by evicting multiple files in one call.
+                let bytes = self.maybe_evict_once().await;
+                if bytes == 0 {
```

Comment:
how about adding a warning log here, if `bytes == 0` while `self.cache_size_bytes.load(Ordering::Relaxed) > target_size ` still true?

it should not happen in the normal cases imo, however if it happened, it should be considered as bug and we need fix it.